### PR TITLE
Close #26150: Do not show sync tabs in home when sync open tabs setting is disabled

### LINKED ISSUE #26150: [Bug]: If tabs are not being Sync'd for an account, the recent synced tab card should not appear
Currently, if a user is logged in to an FxA account but has opted out of syncing tabs, the recent synced tab card appears to load indefinitely on the home screen. Instead, it should just not be displayed.

┆Issue is synchronized with this [Jira Task](https://mozilla-hub.atlassian.net/browse/FNXV2-21094)

### PR DESCRIPTION
**Pull request by rocketsroger**
### Pull Request checklist

- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### QA

- [x] **QA Needed**

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-debug` task.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
Fixes #26150

### CODE REVIEW

**Comment by MatthewTighe** [comment]
@sarah541 FYI, I think this touches some of the files you are also currently working on

**Comment by rocketsroger** [comment]
Seeing some issues while testing.  This PR will need an update before landing.

File: app/src/main/java/org/mozilla/fenix/utils/Settings.kt
Comment (jonalmeida) [inline app/src/main/java/org/mozilla/fenix/utils/Settings.kt:1344]: Shouldn't we be getting this from the Account Manager?

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt:1344]: My original thinking is that this shared preference exists already. Since we update the check box preferences in account settings.  But I'm seeing other issues when turning on and off sync open tabs in account settings.

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt:1344]: Figured out the issue but it's unreleated to this.  I'll use the account manager as suggested.  Thanks

Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: After some investigation, I've moved this check to much earlier so we short circuit if sync tabs is not on.

**Review by jonalmeida — APPROVED** [review]

File: app/src/main/java/org/mozilla/fenix/utils/Settings.kt
Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: What's nice about this change is that it happens in the coroutine flow, so when we check the SyncEnginesStorage, that change happens off the main thread and avoids any perf losses for the homescreen every time the screen is presented.

Comment (jonalmeida) [reply app/src/main/java/org/mozilla/fenix/utils/Settings.kt]: We can also makes it easier to use his [other Synced Tabs change](https://github.com/mozilla-mobile/android-components/pull/12628) when it lands in AC.

**Comment by jonalmeida** [comment]
LGTM, but @MatthewTighe might want to take a quick look at it before it lands as well since he has the most up-to-date context. 🙂

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt:38]: Do you mean to pass in `SyncEnginesStorage(context)` rather than just context?

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt:100]: I'm not sure about this.  This is more of a design change. (ie: do we check before we perform syncs and what type of syncs should we perform).

**Comment by rocketsroger** [comment]
Putting this on hold.  From some discussions we might have a better fix that will fix all the place holder problems.

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt:38]: Yes, that's what I meant. Generally I think `Context` can tend to make unit tests more difficult

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt
Comment (MatthewTighe) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt:132]: Yeah, it seems like this prevents the loading placeholder from appearing at all. It sounded like product was okay with removing the placeholder, but I would make sure to inform design as well. I personally prefer it over having the card pop-in, but I will leave it up to you

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt:68]: Sure, I'll remove the parameter.  Thanks

File: app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt
Comment (rocketsroger) [reply app/src/main/java/org/mozilla/fenix/home/recentsyncedtabs/RecentSyncedTabFeature.kt:132]: I see.  I'll update the solution here so we don't have placeholder for now.  This could be risky if we see the placeholder card pop in and out.  It will make my change smaller and at the same time less risky.

**Review by MatthewTighe — CHANGES_REQUESTED** [review]
This latest push looks like it reintroduces the original issue. If I have a synced tab available and disable tabs syncing in settings then the tab will still be visible when I return to home. It looks like the code to check the engines storage was removed, so the `AppStore` will never be updated with new state.

**Comment by rocketsroger** [comment]
> This latest push looks like it reintroduces the original issue. If I have a synced tab available and disable tabs syncing in settings then the tab will still be visible when I return to home. It looks like the code to check the engines storage was removed, so the `AppStore` will never be updated with new state.

Yes, it stays until the next time the user starts the app.  I was looking for a simple change that remove the problem with blank sync tab card but not more.  Do you think we have to hide the sync tab if we already have it in memory?

**Comment by MatthewTighe** [comment]
> Yes, it stays until the next time the user starts the app. I was looking for a simple change that remove the problem with blank sync tab card but not more. Do you think we have to hide the sync tab if we already have it in memory? 

I think in general if a user disables tab syncing it is probably with the intent that synced tabs are no longer available on their device, so they might be surprised to see it still on their home screen. That case can be handled in a follow-up though

**Comment by rocketsroger** [comment]
> I think in general if a user disables tab syncing it is probably with the intent that synced tabs are no longer available on their device, so they might be surprised to see it still on their home screen. That case can be handled in a follow-up though

Sure, I'll restore the part where we check if the tab sync is off.  Thanks

**Review by MatthewTighe — APPROVED** [review]

**Pull request merged**
